File: android/core/src/main/java/app/visly/shard/viewimpl/ScrollViewImpl.kt
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package app.visly.shard.viewimpl

import android.view.View
import android.widget.HorizontalScrollView
import android.widget.ScrollView
import app.visly.shard.*

class ScrollViewImpl(ctx: ShardContext): BaseViewImpl<View>(ctx) {
    internal enum class Direction {
        VERTICAL,
        HORIZONTAL,
    }

    internal var direction = Direction.VERTICAL
    internal var contentInset = 0
    internal var content: ShardRoot? = null

    override fun measure(width: Float?, height: Float?): Size {
        return content?.measure(width, height) ?: Size(width ?: 0f, height ?: 0f)
    }

    override fun setProp(key: String, value: JsonValue) {
        super.setProp(key, value)

        when (key) {
            "direction" -> {
                direction = when (value) {
                    JsonValue.String("horizontal") -> Direction.HORIZONTAL
                    JsonValue.String("vertical") -> Direction.VERTICAL
                    else -> Direction.VERTICAL
                }
            }

            "content-inset" -> {
                contentInset = when (value) {
                    is JsonValue.Object -> value.toDips(ctx).toInt()
                    else -> 0
                }
            }

            "content" -> {
                content = when (value) {
                    is JsonValue.Object -> ShardViewManager.instance.loadJson(ctx, value)
                    else -> null
                }
            }
        }
    }

    override fun createView(): View {
        val view = if (direction == Direction.VERTICAL) {
            ScrollView(ctx)
        } else {
            HorizontalScrollView(ctx)
        }

        view.clipToPadding = false
        view.isHorizontalScrollBarEnabled = false
        view.isVerticalScrollBarEnabled = false
        view.addView(ShardRootView(ctx))
        return view
    }

    override fun bindView(view: View) {
        super.bindView(view)
        val view = if (direction == Direction.VERTICAL) {
            view as ScrollView
        } else {
            view as HorizontalScrollView
        }

        view.setPadding(
                if (direction == Direction.VERTICAL) 0 else contentInset,
                if (direction == Direction.HORIZONTAL) 0 else contentInset,
                if (direction == Direction.VERTICAL) 0 else contentInset,
                if (direction == Direction.HORIZONTAL) 0 else contentInset)

        val content = content
        if (content != null) {
            val contentRoot = view.getChildAt(0) as ShardRootView
            contentRoot.setRoot(content)
        }
    }
}

File: ios/ShardKit/Classes/ScrollViewImpl.swift
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit

internal enum Direction {
    case vertical
    case horizontal
}

internal class ScrollViewImpl: BaseViewImpl {
    internal var direction: Direction = .vertical
    internal var contentInset = Float(0)
    internal var content: ShardRoot? = nil
    
    override func measure(width: CGFloat?, height: CGFloat?) -> CGSize {
        if let content = content {
            return content.measure(width: width, height: height)
        } else {
            return CGSize(width: width ?? 0, height: height ?? 0)
        }
    }
    
    override func setProp(key: String, value: JsonValue) throws {
        try super.setProp(key: key, value: value)
        
        switch key {
        case "direction":
            switch value {
            case .String(let value) where value == "vertical": self.direction = .vertical
            case .String(let value) where value == "horizontal": self.direction = .horizontal
            default: return self.direction = .vertical
            }
        case "content-inset": self.contentInset = try value.asObject().asDimension()
        case "content":
            let result = ShardViewManager.shared.loadJson(value)
            switch(result) {
            case .Success(let content):
                self.content = content
            case .Failure(let error):
                throw error
            }
        default: ()
        }
    }
    
    override func createView() -> UIView {
        let scroll = UIScrollView()
        let contentRoot = ShardRootView()
        scroll.addSubview(contentRoot)
        return scroll
    }
    
    override func bindView(_ view: UIView) {
        super.bindView(view)
        
        let view = view as! UIScrollView
        view.alwaysBounceVertical = self.direction == .vertical
        view.alwaysBounceHorizontal = self.direction == .horizontal
        
        let inset = CGFloat(self.contentInset)
        if direction == .vertical {
            view.contentInset = UIEdgeInsets(top: inset, left: 0, bottom: inset, right: 0)
            view.contentOffset = CGPoint(x: 0, y: -inset)
        } else {
            view.contentInset = UIEdgeInsets(top: 0, left: inset, bottom: 0, right: inset)
            view.contentOffset = CGPoint(x: -inset, y: 0)
        }
        
        
        if let content = self.content {
            let contentRoot = view.subviews[0] as! ShardRootView
            
            let size = content.layout(
                width: self.direction == .vertical ? view.frame.width : nil,
                height: self.direction == .horizontal ? view.frame.height : nil)
            
            contentRoot.frame = CGRect(origin: .zero, size: size)
            view.contentSize = size
            contentRoot.setRoot(content)
        }
    }
}
